### PR TITLE
Add byKey, byValue/opSlice, and byKeyValue to HashMap

### LINKED ISSUE #75: HashMap should have a byKeyValue Equivalent
Should return a range of tuples with named members

### PR DESCRIPTION
**Pull request by Hackerpilot**
Fixes #75

### CODE REVIEW

File: src/containers/hashmap.d [hunk 1]
```diff
@@ -85,12 +85,26 @@ struct HashMap(K, V, Allocator = Mallocator, alias hashFunction = generateHash!K
...
 	}
 
-	~this()
+	~this() nothrow
 	{
-		import stdx.allocator : dispose;
+		try
+			clear();
+		catch (Throwable)
+			assert(false);
```
Comment (wilzbach) [inline src/containers/hashmap.d]: `scope(failure) assert (0);` is the idiomatic way though that's nit picking

File: src/containers/hashmap.d [hunk 2]
```diff
@@ -85,12 +85,26 @@ struct HashMap(K, V, Allocator = Mallocator, alias hashFunction = generateHash!K
...
+	}
+
+	/**
+	 * Removes all items from the map
+	 */
+	void clear()
+	{
+		import std.experimental.allocator : dispose;
```
Comment (wilzbach) [inline src/containers/hashmap.d]: What's the reason for not using stdx.allocator? The experimental package may break at any time again :/

Comment (Hackerpilot) [reply src/containers/hashmap.d]: That's something left over from when I was working on getting this working. Maybe the allocators change broke something else downstream, but the containers library itself seemed to work fine.

File: src/containers/hashmap.d (hunk 1 above)
Comment (Hackerpilot) [reply src/containers/hashmap.d]: Makes sense, but this is the first time I've seen that suggested.

**Pull request merged**